src/transport/concentration_model.cc

```diff
@@ -321,36 +326,42 @@ void ConcentrationTransportModel::get_flux_bc_sigma(unsigned int index,
 
 void ConcentrationTransportModel::compute_source_coefficients(const std::vector<arma::vec3> &point_list,
 			const ElementAccessor<3> &ele_acc,
-			std::vector<arma::vec> &sources_value,
-			std::vector<arma::vec> &sources_density,
-			std::vector<arma::vec> &sources_sigma)
+			std::vector<std::vector<double> > &sources_value,
+			std::vector<std::vector<double> > &sources_density,
+			std::vector<std::vector<double> > &sources_sigma)
 {
 	const unsigned int qsize = point_list.size();
 	vector<double> csection(qsize);
 	data().cross_section.value_list(point_list, ele_acc, csection);
-	data().sources_conc.value_list(point_list, ele_acc, sources_value);
-	data().sources_density.value_list(point_list, ele_acc, sources_density);
-	data().sources_sigma.value_list(point_list, ele_acc, sources_sigma);
-
-	for (unsigned int k=0; k<qsize; k++)
-	{
-		sources_density[k] *= csection[k];
-		sources_sigma[k] *= csection[k];
-	}
+    for (unsigned int sbi=0; sbi<n_substances(); sbi++)
+    {
+      data().sources_conc[sbi].value_list(point_list, ele_acc, sources_value[sbi]);
+      data().sources_density[sbi].value_list(point_list, ele_acc, sources_density[sbi]);
+      data().sources_sigma[sbi].value_list(point_list, ele_acc, sources_sigma[sbi]);
+      
+      for (unsigned int k=0; k<qsize; k++)
+      {
+          sources_density[sbi][k] *= csection[k];
+          sources_sigma[sbi][k] *= csection[k];
+      }
+    }
 }
 
 
 void ConcentrationTransportModel::compute_sources_sigma(const std::vector<arma::vec3> &point_list,
```

Comment:
Unrelated question: Why some functions have point_list and some not?
Is there repeated evaluation of some fields? Possible optimization by field caches ... Can you create an issue for that, with list of places to consider?

Reply:
Which functions do not have point_list?
All methods in AdvectionDiffusionModel take point_list, if I am not mistaken.